keps/sig-node/2371-cri-pod-container-stats/README.md

```diff
@@ -0,0 +1,730 @@
+
+### Notes/Constraints/Caveats (Optional)
+
+#### History/Past Conversations
+
+There have been conversations in the past on [deprecating the stats Summary API][#core-proposal], in favor of allowing an on-demand model of stats gathering/reporting.
+This KEP does not attempt to implement that. These stats are largely considered [a part of Kubernetes][#keep-summary]. The stated goal of this KEP is to reduce the overhead
+on two entities reporting metrics, not totally changing what stats the Kubelet reports (with the exception of Accelerator and UserDefinedMetrics).
+
+Thus, this KEP largely the plan described [here][#plan], with some changes:
+
+- the `/metrics/cadvisor` endpoint will likely not be exposed in the Alpha release, and its continued support will need to be evaluated for the future.
+- CRI is used for all of the monitoring endpoints related to Containers and Pods (except Volume and Ephemeral Storage)
+- CRI is used to provide metrics for eviction.
+
+[core-proposal]: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/core-metrics-pipeline.md#proposed-core-metrics
+[keep-summary]: https://github.com/kubernetes/kubernetes/issues/68522#issuecomment-666636130
+[plan]: https://github.com/kubernetes/kubernetes/issues/68522#issuecomment-724928827
+
+#### cAdvisor Metrics Endpoint
+
+In addition to the `/stats/summary` endpoint, there are also consumers that rely on `/metrics/cadvisor`.
+In the past, each of these were populated primarily by cAdvisor, but the goals of this KEP will prevent cAdvisor from hosting these metrics.
+Where the majority of internal Kubernetes entities rely solely on the Summary API, there are likely consumers in the community that rely on this endppoint.
+
+Thus, in addition to fulfilling `/stats/summary` endpoint, this KEP will have to decide on how to handle `/metrics/cadvisor`
+Below is the proposed strategy for doing so:
+
+1. For the Alpha release of this KEP, accept there will be breaking changes to `/metrics/cadvisor`. Use a feature gate and the strategies proposed above to minimize breakages.
+2. For the Beta release, evaluate the future of the `/metrics/cadvisor` endpoint. We can either:
+    - Deprecate the endpoint, and extend the Summary API to report the required fields.
+    - Move the reporter of `/metrics/cadvisor` from cAdvisor to the CRI implementation.
+    - Extend the CRI and have the CRI implementation report the required stats, and have the Kubelet implement the `/metrics/cadvisor` endpoint. 
+3. For the GA release, the CRI implementation should be the source of truth for all pod and container level metrics that external parties rely on (no matter how many endpoints the Kubelet advertises).
+
+
+### Risks and Mitigations
+
+- To properly move to CRI stats, it is likely there are some metrics that we'll want to not support (Accelerator/UserDefined). We should be careful to not break entities that rely on these metrics.
```

Comment:
Is it possible for the container runtimes to emit those stats in some other way? I know the table discussed CAdvisor as a daemonset, are there other alternatives?

Reply:
For Accelerator, I think we're safe to not do any work supporting them.

UserDefinedMetrics is a bit trickier. It definitely could be reported by CRI implementations, but I think it's best suited as an additional entity (daemonset or other) that emits it to prometheus. That seems to be the more sustainable (and idiomatic) solution.

@dashpole do you have thoughts here

Reply:
For accelerator metrics, you could reference the nvidia [dcgm-exporter](https://github.com/NVIDIA/gpu-monitoring-tools#dcgm-exporter) if you wanted as an alternative.  The accelerator deprecation KEP has a table that shows the relationship between existing metrics and the dcgm ones IIRC.

@ehashman is your question about metrics we are dropping from /stats/summary, or metrics from /metrics/cadvisor?  I don't think the /stat/summary differences will be an issue, but the /metrics/cadvisor metrics we would drop would probably be more disruptive to users.  I don't know of other ways to get many of those metrics.